gsy_framework/sim_results/electric_blue/all_results.py

```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from typing import Dict
+
+from pendulum import DateTime, Duration
+
+from gsy_framework.enums import AvailableMarketTypes
+from gsy_framework.forward_markets.forward_profile import ALLOWED_MARKET_TYPES
+from gsy_framework.sim_results.electric_blue.aggregate_results import (handle_forward_results,
+                                                                       ForwardDeviceStats,
+                                                                       MARKET_RESOLUTIONS)
+from gsy_framework.sim_results.electric_blue.time_series import ForwardDeviceTimeSeries
+from gsy_framework.utils import str_to_pendulum_datetime
+
+
+class ForwardResultsHandler:
+    """Calculate all results for each market slot for forward markets."""
+
+    def __init__(self, slot_length: Duration):
+        self.forward_market_enabled = True
+        self.slot_length = slot_length
+        self.orders: Dict[
+            str, Dict[int, Dict[DateTime, Dict]]] = defaultdict(lambda: defaultdict(dict))
+        self.previous_device_stats = {}
+        self.current_device_stats: Dict[
+            int, Dict[DateTime, Dict[str, Dict]]] = defaultdict(lambda: defaultdict(dict))
+        self.device_time_series: Dict[
+            int, Dict[int, Dict[DateTime, Dict[str, Dict]]]] = defaultdict(
+            lambda: defaultdict(lambda: defaultdict(dict)))
+        self._total_memory_utilization_kb = 0.0
+
+    def update(self, _area_dict: Dict, core_stats: Dict, current_market_slot: str) -> None:
+        """
+        Update the forward market aggregated results with bids_offers_trades of current slot.
+        """
+        self._clear_device_statistics()
+        for area_uuid, area_result in core_stats.items():
+            if "forward_market_stats" not in area_result:
+                return
+
+            current_market_dt = str_to_pendulum_datetime(current_market_slot) if (
+                current_market_slot) else ""
```

Comment:
the name already suggests, that it should be a DateTime object. So, writing an empty string to it in the else case, will not work. So, `current_market_slot` can be None, right? Can you please give a hint in the interface of the method: `Optional[str]`?  When is the current_market_slot empty? In the very first send of the results?
How should `handle_forward_results` behave in this case? It expects a DateTime object, right?

Reply:
Two first messages always have `current_market_slot` as an empty string and slot number 0 and usually no `simulation_raw_data` (That's why `handle_forward_results` had no issue with this, since there was nothing to iterate through). Then the third message is 'T00:00' and slot number is 1 and starts to have data in `simulation_raw_data`. I think that we should filter out these entries with empty string as `current_market_slot` at the very beginning of processing. I will fix it, thanks!